Sources/SwiftParser/Declarations.swift

```diff
@@ -1041,6 +1041,14 @@ extension Parser {
     // Parse the signature.
     let signature = self.parseFunctionSignature()
 
+    // mark the unexpected result type of initializer
+    // and replace the ‘output‘ of signature with nil
+    var unexpectedResultType: RawUnexpectedNodesSyntax?
+    if let output = signature.output {
+      signature.raw = signature.layoutView.replacingChild(at: 5, with: nil, arena: self.arena)
```

Comment:
Replacing a child by its index is super fragile because the index might change if a new child is added to `RawFunctionSignatureSyntax`. What I would prefer is to add a parameter to `parseFunctionSignature` like `allowOutput` that, if set to `false` just doesn’t parse an output type in the function signature.

Reply:
Hey @ahoppen , I've just made a change about this.

Instead of just not parsing the output type, I make the output nil and create an unexpected node in ``parseFunctionSignature`` (not using ``replacingChild``), which is controlled by the flag ``allowOutput``.

If I leave the output type unparsed, there will be some corner cases, e.g. 
```
init(_ foo: T) -> Int where T: Comparable {}
```

in which the 'where' clause will be left unparsed and recognized as a part of the unexpected node, which is not.  I've also added this snippet as a new testcase.

BTW, if I promote the recovery precedence of ``where`` and try to recover to it will bring other problems.

I wonder whether this change is reasonable? Thanks in advance!